adapters/adkernelAdn/adkernelAdn.go

```
@@ -0,0 +1,242 @@
+	if len(*imps) == 0 {
+		return nil, *err
+	}
+	errs = append(errs, *err...)
+
+	pub2impressions := dispatchImpressions(imps, impExts)
+	result := make([]*adapters.RequestData, 0, len(pub2impressions))
+	for k, imps := range pub2impressions {
+		bidRequest, err := buildAdapterRequest(request, &k, &imps)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			result = append(result, bidRequest)
+		}
+	}
+	return result, errs
+}
+
+// getImpressionsInfo checks each impression for validity and returns impressions copy with corresponding exts
+func getImpressionsInfo(imps *[]openrtb.Imp) (*[]openrtb.Imp, *[]openrtb_ext.ExtImpAdkernelAdn, *[]error) {
+	impsCount := len(*imps)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb.Imp, 0, impsCount)
+	resImpExts := make([]openrtb_ext.ExtImpAdkernelAdn, 0, impsCount)
+
+	for _, imp := range *imps {
+		impExt, err := getImpressionExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		if err := validateImpression(&imp, impExt); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		resImps = append(resImps, imp)
+		resImpExts = append(resImpExts, *impExt)
+
+	}
+	return &resImps, &resImpExts, &errors
+}
+
+func validateImpression(imp *openrtb.Imp, impExt *openrtb_ext.ExtImpAdkernelAdn) error {
+	if imp.Banner == nil && imp.Video == nil {
+		return newBadInputError(fmt.Sprintf("Only banner and video impressions are supported. Ignoring imp id=%s", imp.ID))
+	}
+	if impExt.PublisherID < 1 {
+		return newBadInputError(fmt.Sprintf("Invalid pubId value. Ignoring imp id=%s", imp.ID))
+	}
+	return nil
+}
+
+//Group impressions by AdKernel-specific parameters `pubId` & `host`
+func dispatchImpressions(imps *[]openrtb.Imp, impsExt *[]openrtb_ext.ExtImpAdkernelAdn) map[openrtb_ext.ExtImpAdkernelAdn][]openrtb.Imp {
+	res := make(map[openrtb_ext.ExtImpAdkernelAdn][]openrtb.Imp)
+	for idx := range *imps {
+		imp := (*imps)[idx]
+		compatImpression(&imp)
```

You are ignoring any errors returned by `compatImpression()` currently?

---

Looks like my mistake. Thank you

---

Addressed in 076145eba887ea2ac2eab38ffaef77e9a9ef631c